sdk/core/System.ClientModel/src/Message/BinaryContent.cs

```diff
@@ -0,0 +1,179 @@
+using System.ClientModel.Primitives;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.ClientModel;
+
+public abstract class BinaryContent : IDisposable
+{
+    private static readonly ModelReaderWriterOptions ModelWriteWireOptions = new ModelReaderWriterOptions("W");
+
+    /// <summary>
+    /// Creates an instance of <see cref="BinaryContent"/> that wraps a <see cref="BinaryData"/>.
+    /// </summary>
+    /// <param name="value">The <see cref="BinaryData"/> to use.</param>
+    /// <returns>An instance of <see cref="BinaryContent"/> that wraps a <see cref="BinaryData"/>.</returns>
+    public static BinaryContent Create(BinaryData value) => new BinaryDataBinaryContent(value.ToMemory());
+
+    /// <summary>
+    /// Creates an instance of <see cref="BinaryContent"/> that wraps a <see cref="IPersistableModel{T}"/>.
+    /// </summary>
+    /// <param name="model">The <see cref="IPersistableModel{T}"/> to write.</param>
+    /// <param name="options">The <see cref="ModelReaderWriterOptions"/> to use.</param>
+    /// <returns>An instance of <see cref="BinaryContent"/> that wraps a <see cref="IPersistableModel{T}"/>.</returns>
+    public static BinaryContent Create<T>(T model, ModelReaderWriterOptions? options = default) where T : IPersistableModel<T>
+        => new ModelBinaryContent<T>(model, options ?? ModelWriteWireOptions);
+
+    /// <summary>
+    /// Attempts to compute the length of the underlying body content, if available.
+    /// </summary>
+    /// <param name="length">The length of the underlying data.</param>
+    public abstract bool TryComputeLength(out long length);
+
+    /// <summary>
+    /// Writes contents of this object to an instance of <see cref="Stream"/>.
+    /// </summary>
+    /// <param name="stream">The stream to write to.</param>
+    /// <param name="cancellationToken">To cancellation token to use.</param>
+    public abstract Task WriteToAsync(Stream stream, CancellationToken cancellationToken);
+
+    /// <summary>
+    /// Writes contents of this object to an instance of <see cref="Stream"/>.
+    /// </summary>
+    /// <param name="stream">The stream to write to.</param>
+    /// <param name="cancellationToken">To cancellation token to use.</param>
+    public abstract void WriteTo(Stream stream, CancellationToken cancellationToken);
+
+    /// <inheritdoc/>
+    public abstract void Dispose();
+
+    private sealed class BinaryDataBinaryContent : BinaryContent
+    {
+        private readonly ReadOnlyMemory<byte> _bytes;
+
+        public BinaryDataBinaryContent(ReadOnlyMemory<byte> bytes)
+        {
+            _bytes = bytes;
+        }
+
+        public override bool TryComputeLength(out long length)
+        {
+            length = _bytes.Length;
+            return true;
+        }
+
+        public override void WriteTo(Stream stream, CancellationToken cancellation)
+        {
+            byte[] buffer = _bytes.ToArray();
```

Comment:
Should we do an `#if NET6_0_OR_GREATER` here and use the `Write(ReadOnlySpan<Byte>)` overload to avoid the array allocation where we can?

I'm thinking something like:
```csharp
#if NET6_0_OR_GREATER
    stream.Write(_bytes.Span);
#else
    byte[] buffer = _bytes.ToArray();
    stream.Write(buffer, 0, buffer.Length);
#endif
```

Reply:
Looks like we do this on L151 already.